### PR TITLE
Authorization Code with PKCE in Single-Page Applications: Proof-of-Concept

### PR DESCRIPTION
**Pull request by pkanher617**
This is a Proof of Concept for the using the OAuth 2.0 Authorization Code Flow with PKCE in Single-Page Applications described in issue #1000. This PR is the first of many to continually improve the MSAL library and bring this flow to parity. 

As a part of this POC, we have also separated much of the platform-specific code from the protocol code, and created separate packages. This will enable us to use this flow for other platforms in the future, including Electron, Node.js, and more.

This PR should eventually have a working version of the POC for Electron using the Auth Code implementation in the msal-common project.

### CODE REVIEW

File: lib/msal-browser-poc/rollup.config.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import typescript from "rollup-plugin-typescript2";
+import { uglify } from "rollup-plugin-uglify";
+import resolve from "rollup-plugin-node-resolve";
+import pkg from "./package.json";
+
+const libraryHeader = `/*! ${pkg.name} v${pkg.version} ${new Date().toISOString().split('T')[0]} */`;
+const useStrictHeader = `'use strict';`;
+const fileHeader = `${libraryHeader}\n${useStrictHeader}`;
+
+export default [
```
Comment (DarylThayil) [inline lib/msal-browser-poc/rollup.config.js]: looking forward to hearing / reading your thoughts on why roll up may be a good choice for the lib

File: lib/msal-browser-poc/src/app/PublicClientApplication.ts [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        this.config = buildConfiguration(configuration);
+
+        // Create browser storage
+        this.cacheStorage = new BrowserStorage(this.config.auth.clientId, this.config.cache);
+
+        // Initialize the network module
+        this.networkClient = new FetchClient();
+
+        // Initialize crypto module
+        this.crypto = new BrowserCrypto();
```
Comment (DarylThayil) [inline lib/msal-browser-poc/src/app/PublicClientApplication.ts]: do any of these have to be set on the instance can they just be passed along ?

File: lib/msal-browser-poc/src/error/ClientBrowserAuthError.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/*
```
Comment (DarylThayil) [inline lib/msal-browser-poc/src/error/ClientBrowserAuthError.ts]: I wonder if they errors need to be split by browser or non browser.

for instance a common/errors/ConfigurationError, could be used by both and then browser specific classes might life in msal-browser such as IframeError (or something)

File: lib/msal-browser-poc/src/network/XHRClient.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * XHR client for JSON endpoints
+ * https://www.npmjs.com/package/async-promise
+ * @hidden
+ */
+export class XhrClient implements IXhrClient {
+    public sendRequestAsync(url: string, requestParams: RequestInit, enableCaching?: boolean): Promise<any> {
+        return new Promise<string>((resolve, reject) => {
+            const xhr = new XMLHttpRequest();
```
Comment (DarylThayil) [inline lib/msal-browser-poc/src/network/XHRClient.ts]: can we use fetch?

File: lib/msal-common-poc/src/app/MsalConfiguration.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+ *
+ * @param TAuthOptions
+ * @param TStorageOptions
+ * @param TSystemOptions
+ * @param TFrameworkOptions
+ *
+ * @returns TConfiguration object
+ */
+export function buildConfiguration({ auth, storageInterface, networkInterface, cryptoInterface }: MsalConfiguration): MsalConfiguration {
+    const overlayedConfig: MsalConfiguration = {
```
Comment (DarylThayil) [inline lib/msal-common-poc/src/app/MsalConfiguration.ts]: do we think configuration between server and browser will use the same common object?

File: lib/msal-browser-poc/package.json [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    "directories": {
+        "test": "test"
+    },
+    "files": [
+        "dist"
+    ],
+    "scripts": {
+        "clean": "shx rm -rf dist lib",
+        "lint": "eslint src --ext .ts",
+        "build:all": "npm run build:common && npm link msal-common && npm run build",
```
Comment (jasonnutter) [inline lib/msal-browser-poc/package.json]: `lerna bootstrap` will take care of the linking for you.

File: lib/msal-browser-poc/package.json [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        "@typescript-eslint/parser": "^2.4.0",
+        "eslint": "^6.5.1",
+        "husky": "^3.0.9",
+        "rollup": "^1.24.0",
+        "rollup-plugin-node-resolve": "^5.2.0",
+        "rollup-plugin-typescript2": "^0.24.3",
+        "rollup-plugin-uglify": "^6.0.3",
+        "shx": "^0.3.2",
+        "tslib": "^1.10.0",
+        "tslint": "^5.20.0",
```
Comment (jasonnutter) [inline lib/msal-browser-poc/package.json]: Thought this needs to be a normal dep?

File: lib/msal-browser-poc/package.json [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        "js",
+        "AAD",
+        "msal",
+        "oauth"
+    ],
+    "main": "./dist/index.js",
+    "module": "./dist/index.es.js",
+    "types": "./dist/index.d.ts",
+    "engines": {
+        "node": ">=0.8.0"
```
Comment (jasonnutter) [inline lib/msal-browser-poc/package.json]: We should change this to whatever version is run in CI

File: lib/msal-browser-poc/package.json [hunk 9]
```diff
@@ -0,0 +1,62 @@
+{
+    "name": "msal-browser",
```
Comment (jasonnutter) [inline lib/msal-browser-poc/package.json]: We need to decide if we want this or `@azure/msal-browser` (which is probably better for branding/consistency)

File: lib/msal-browser-poc/src/network/XHRClient.ts (hunk 4 above)
Comment (pkanher617) [reply lib/msal-browser-poc/src/network/XHRClient.ts]: This implementation is for the implicit flow, in the auth code library we will switch to using fetch.

File: lib/msal-common-poc/src/app/MsalConfiguration.ts (hunk 5 above)
Comment (pkanher617) [reply lib/msal-common-poc/src/app/MsalConfiguration.ts]: I think we will have different auth options for server and browser, but everything else will be the same. So I am separating that into two different config objects in PR #1109.

File: lib/msal-browser-poc/package.json (hunk 8 above)
Comment (pkanher617) [reply lib/msal-browser-poc/package.json]: Ok is there a recommendation for what version we should be using? In CI we are just using latest stable.

File: lib/msal-browser-poc/package.json (hunk 7 above)
Comment (pkanher617) [reply lib/msal-browser-poc/package.json]: You're correct, I will rectify this in PR #1109.

File: lib/msal-browser-poc/package.json (hunk 6 above)
Comment (pkanher617) [reply lib/msal-browser-poc/package.json]: I will implement lerna in a future PR, just trying to get POC working for now.

File: lib/msal-browser-poc/package.json (hunk 9 above)
Comment (pkanher617) [reply lib/msal-browser-poc/package.json]: We should decide on this in PR #1108.

File: lib/msal-browser-poc/src/error/ClientBrowserAuthError.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-browser-poc/src/error/ClientBrowserAuthError.ts]: That's the idea, I will flesh this out more in future PRs.

**Pull request closed without merging**